tests/kernel/smp/src/main.c

```diff
@@ -37,45 +52,354 @@ void t2_fn(void *a, void *b, void *c)
+		 * count will always be lower
+		 */
+		k_busy_wait(DELAY_US + (DELAY_US / 8));
+
+		zassert_true(t2_count > i, "T2 thread isn't"
+			     " scheduled to run on other core");
+	}
+	/* Abort the t2_fn thread */
+	k_thread_abort(tid);
+
+	/* At this point of time, 1 core has current thread running,
+	 * and other core has test_main thread running.
+	 * Once current thread is terminated, test_main reports the
+	 * test suite execution pass/fail.
 	 */
-	k_sleep(100);
+}
 
-	/* Create a thread at a fixed priority lower than the main
-	 * thread.  In uniprocessor mode, this thread will never be
-	 * scheduled and the test will fail.
+/**
+ * @brief Verify multi core execution with 2 cooperative threads
+ *
+ * @details Create a cooperative thread, since current thread is
+ * also cooperative, there is no chance that this thread
+ * will get scheduled in uniprocessor mode.
+ *
+ * For ESP32 only (has only 2 cores): Since t2 thread has
+ * higher priority than test_main and test_main is preemptive
+ * test_main should be moved to timeout queue and t2
+ * thread should get scheduled instead. So t2 thread should
+ * be in one core and test_smp_coop_threads should be
+ * another core.
+ */
+void test_smp_coop_threads(void)
+{
+	/* In a multi processor mode, this thread should have
+	 * been scheduled on another core other than the one
+	 * which has test_main scheduled
 	 */
-	k_thread_create(&t2, t2_stack, T2_STACK_SIZE, t2_fn,
-			NULL, NULL, NULL,
-			CONFIG_MAIN_THREAD_PRIORITY + 1, 0, K_NO_WAIT);
+	int i;
+
+	k_tid_t tid = k_thread_create(&t2, t2_stack, STACK_SIZE, t2_fn,
+				      NULL, NULL, NULL,
+				      K_PRIO_COOP(1), 0, K_NO_WAIT);
 
 	for (i = 0; i < 10; i++) {
-		/* Wait slightly longer than the other thread so our
-		 * count will always be lower
-		 */
 		k_busy_wait(DELAY_US + (DELAY_US / 8));
 
-		if (t2_count <= i) {
-			ok = 0;
-			break;
-		}
+		zassert_true(t2_count > i, "Another thread isn't"
+			     " scheduled to run on other core");
+	}
+	k_thread_abort(tid);
+}
+
+static void smp_thread_entry(void *p1, void *p2, void *p3)
+{
+	int tnum = (int)p1;
+
+	TC_PRINT("Thread started\n");
+	k_busy_wait(THREAD_DELAY);
+
+	tdata[tnum].executed = 1;
+	TC_PRINT("Thread execution complete\n");
+}
+
+/**
+ * @brief Verify round robin on one core with coop thread
+ *
+ * @details The cooperative test creates THREAD_NUM preemptive
+ * threads of same priority and checks the round robin behavior.
+ * the cooperative thread will be blocking one core hence
+ * preemptive threads should run on another core in round robin.
+ */
+static void test_smp_single_core_coop_rr_threads(void)
+{
+	int i;
+
+	k_sched_time_slice_set(100, 0);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_LOWEST_THREAD_PRIO - 1),
```

Comment:
No point in specifying both K_PRIO_PREEMPT and K_LOWEST_THREAD_PRIORITY, the concepts basically mean the same thing.   Also there's a K_LOWEST_APPLICATION_THREAD_PRIO symbol that already means what you want.

(Yes, our priority selection API is sort of a mess.)